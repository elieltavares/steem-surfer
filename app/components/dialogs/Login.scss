.login-dialog-content {
  .login-msg {
    margin-bottom: 10px;
  }

  .account-list {
    .account-list-header {
      font-size: 18px;
      font-weight: 500;
      margin-bottom: 10px;
      margin-top: 10px;
      text-align: center;

      @include themify(day) {
        color: $steel-grey;
      }

      @include themify(night) {
        color: $metallic-blue;
      }
    }

    .account-list-body {
      .account-list-item {
        align-items: center;
        border-radius: 25px;
        cursor: pointer;
        display: flex;
        font-size: 16px;
        line-height: 16px;
        margin-bottom: 10px;
        padding: 10px;
        position: relative;
        transition: 0.5s;

        @include themify(day) {
          background: $duck-egg-blue;
          color: $dark-sky-blue;
        }

        @include themify(night) {
          background: $metallic-blue;
          color: $powder-blue;
        }

        &:hover {
          @include themify(day) {
            background: lighten($dark-sky-blue, 30);
          }

          @include themify(night) {
            background: darken($metallic-blue, 16);
          }
        }

        .user-avatar {
          margin-right: 10px;
        }

        .username {
          font-weight: 500;
        }

        .space {
          flex-grow: 1;
        }

        .check-mark {
          background: $dark-sky-blue;
          border-radius: 50%;
          height: 5px;
          left: -14px;
          position: absolute;
          top: calc(50% - 2.5px);
          width: 5px;
        }

        .btn-delete {
          align-items: center;
          border-radius: 50%;
          cursor: pointer;
          display: none;
          height: 24px;
          justify-content: center;
          width: 24px;

          .mi {
            color: $white;
            font-size: 16px;
          }

          &:hover {
            .mi {
              color: lighten($red2, 10);
            }
          }

          @include themify(day) {
            background: darken($duck-egg-blue, 20);

            &:hover {
              background: $duck-egg-blue;
            }
          }

          @include themify(night) {
            background: darken($metallic-blue, 3);

            &:hover {
              background: $metallic-blue;
            }
          }
        }

        &:hover {
          .btn-delete {
            display: flex;
          }
        }
      }
    }
  }

  .login-sc {
    cursor: pointer;
    margin: auto;
    text-align: center;

    .login-title {
      font-weight: 700;
      margin: 0 0 10px 0;
    }

    .logo {
      background: $white-two;
      border-radius: 4px;
      margin: 0 auto 14px auto;
      padding: 3px;
      width: 190px;
    }

    .sc-note {
      font-size: 95%;
      margin-bottom: 10px;
      opacity: 0.5;
    }
  }
}

.login-dialog-content,
.sign-up-dialog-content {
  margin: auto;
  padding: 20px;
  width: 80%;

  .dialog-header {
    .logo {
      margin: 0 auto 10px auto;
      width: 100px;

      img {
        width: 100%;
      }
    }

    .dialog-header-text {
      font-size: 22px;
      font-weight: 700;
      margin-bottom: 10px;
      text-align: center;
    }
  }

  .dialog-form {
    p {
      color: $steel-grey;
      margin-bottom: 14px;
      text-align: center;
    }
  }
}

.sign-up-dialog-content {
  .dialog-header {
    .dialog-header-text {
      margin-bottom: 20px;
    }
  }

  .sign-up-success {
    font-size: 16px;

    p {
      display: flex;
      align-items: center;
      justify-content: center;

      .mi {
        font-size: 24px;
        margin-right: 10px;
        color: $green1;
      }
    }
  }
}
