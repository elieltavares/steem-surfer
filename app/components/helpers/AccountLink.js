import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { getAccount } from '../../backend/steem-client';

export const makePath = username => `/@${username}`;

class AccountLink extends Component {
  goProfile = async () => {
    const { username, history, actions, onClick, afterClick } = this.props;

    onClick();

    let { accountData } = this.props;
    const { setVisitingAccount } = actions;

    if (!accountData) {
      try {
        accountData = await getAccount(username);
      } catch (err) {
        accountData = null;
      }
    }

    if (accountData) {
      setVisitingAccount(accountData);
    }

    history.push(makePath(username));

    afterClick();
  };

  render() {
    const { children } = this.props;

    return React.cloneElement(children, {
      onClick: this.goProfile
    });
  }
}

AccountLink.defaultProps = {
  accountData: null,
  onClick: () => {},
  afterClick: () => {}
};

AccountLink.propTypes = {
  children: PropTypes.element.isRequired,
  username: PropTypes.string.isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  accountData: PropTypes.instanceOf(Object),
  onClick: PropTypes.func,
  afterClick: PropTypes.func,
  actions: PropTypes.shape({
    setVisitingAccount: PropTypes.func.isRequired
  }).isRequired
};

export default AccountLink;
