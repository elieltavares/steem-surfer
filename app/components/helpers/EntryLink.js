import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { getContent } from '../../backend/steem-client';

export const makePath = (category, author, permlink, toReplies = false) =>
  `/${category}/@${author}/${permlink}${toReplies ? '#replies' : ''}`;

class EntryLink extends Component {
  goEntry = async () => {
    let { entry } = this.props;
    const { history, actions, toReplies, markAsRead } = this.props;

    if (markAsRead) {
      markAsRead();
    }

    if (!entry) {
      const { author, permlink } = this.props;
      entry = await getContent(author, permlink);
    }

    if (entry) {
      const { category, author, permlink } = entry;
      actions.setVisitingEntry(entry);
      history.push(makePath(category, author, permlink, toReplies));
    }
  };

  render() {
    const { children } = this.props;

    return React.cloneElement(children, {
      onClick: this.goEntry
    });
  }
}

EntryLink.defaultProps = {
  entry: null,
  toReplies: false,
  markAsRead: null
};

EntryLink.propTypes = {
  children: PropTypes.element.isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  entry: PropTypes.shape({
    category: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    permlink: PropTypes.string.isRequired
  }),
  author: PropTypes.string.isRequired,
  permlink: PropTypes.string.isRequired,
  toReplies: PropTypes.bool,
  actions: PropTypes.shape({
    setVisitingEntry: PropTypes.func.isRequired
  }).isRequired,
  markAsRead: PropTypes.func
};

export default EntryLink;
