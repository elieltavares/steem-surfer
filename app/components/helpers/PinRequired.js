import React, { Component, Fragment } from 'react';
import { Modal } from 'antd';

import PropTypes from 'prop-types';
import PinConfirm from '../dialogs/PinConfirm';

class PinRequired extends Component {
  constructor(props) {
    super(props);

    this.state = {
      show: false
    };
  }

  onPinInvalidated = () => {
    this.setState({ show: false });
  };

  onConfirmPinSuccess = value => {
    const { onSuccess } = this.props;
    this.setState({ show: false }, () => {
      onSuccess(value);
    });
  };

  cancelFn = () => {
    this.setState({ show: false });
  };

  render() {
    const { children } = this.props;
    const { show } = this.state;

    // Clone child element rewriting onclick event
    const newChild = React.cloneElement(children, {
      onClick: () => {
        this.setState({ show: true });
      }
    });

    return (
      <Fragment>
        {show && (
          <Modal
            footer={null}
            closable
            onCancel={this.cancelFn}
            keyboard
            visible
            width="500px"
            centered
            destroyOnClose
          >
            <PinConfirm
              {...this.props}
              onSuccess={this.onConfirmPinSuccess}
              onInvalidate={this.onPinInvalidated}
            />
          </Modal>
        )}
        {newChild}
      </Fragment>
    );
  }
}

PinRequired.defaultProps = {
  onSuccess: () => {}
};

PinRequired.propTypes = {
  children: PropTypes.element.isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  onSuccess: PropTypes.func
};

export default PinRequired;
