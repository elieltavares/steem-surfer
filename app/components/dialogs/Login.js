/*
eslint-disable react/no-multi-comp, jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events
*/

import React, { Component, Fragment } from 'react';

import PropTypes from 'prop-types';

import { PrivateKey, PublicKey } from 'dsteem';

import {
  Button,
  Divider,
  Input,
  Alert,
  message,
  Popconfirm,
  Modal
} from 'antd';

import { FormattedMessage } from 'react-intl';

import Tooltip from '../common/Tooltip';

import UserAvatar from '../elements/UserAvatar';

import PinConfirm from './PinConfirm';

import { getAccounts } from '../../backend/steem-client';

import {
  scTokenRenew,
  usrActivity,
  signUp as signUpFn
} from '../../backend/esteem-client';

import { scLogin } from '../../helpers/sc';

import scLogo from '../../img/steem-connect.svg';

import logo from '../../img/logo-big.png';

class AccountItem extends Component {
  constructor(props) {
    super(props);

    this.state = {
      del: false
    };
  }

  deleteAccount = () => {
    const { activeAccount, actions, account } = this.props;
    const { username } = account;

    if (activeAccount && activeAccount.username === username) {
      actions.logOut();
    }

    actions.deleteAccount(username);

    setTimeout(() => {
      window.dispatchEvent(
        new CustomEvent('account-deleted', { detail: { username } })
      );
    }, 300);
  };

  render() {
    const { account, intl, activeAccount, onSelect } = this.props;
    const { del } = this.state;

    const activeUsername = activeAccount ? activeAccount.username : null;

    return (
      <div
        key={account.username}
        className={`account-list-item ${
          activeUsername === account.username ? 'active' : ''
        }`}
        onClick={() => {
          onSelect(account.username);
        }}
        role="none"
      >
        <UserAvatar {...this.props} user={account.username} size="normal" />{' '}
        <span className="username">@{account.username}</span>
        {activeUsername === account.username && <div className="check-mark" />}
        <div className="space" />
        <Popconfirm
          title={intl.formatMessage({ id: 'g.are-you-sure' })}
          okText={intl.formatMessage({ id: 'g.ok' })}
          cancelText={intl.formatMessage({ id: 'g.cancel' })}
          onCancel={e => {
            e.stopPropagation();
            this.setState({ del: false });
          }}
          onConfirm={e => {
            e.stopPropagation();
            this.setState({ del: true });
          }}
        >
          <div
            role="none"
            className="btn-delete"
            onClick={e => {
              e.stopPropagation();
            }}
          >
            <Tooltip title={intl.formatMessage({ id: 'g.delete' })}>
              <i className="mi">delete_forever</i>
            </Tooltip>
          </div>
        </Popconfirm>
        {del && (
          <Modal
            footer={null}
            closable
            onCancel={e => {
              this.setState({ del: false });
              e.stopPropagation();
            }}
            keyboard
            visible
            width="500px"
            centered
            destroyOnClose
            maskClosable={false}
          >
            <PinConfirm {...this.props} onSuccess={this.deleteAccount} />
          </Modal>
        )}
      </div>
    );
  }
}

AccountItem.defaultProps = {
  activeAccount: null
};

AccountItem.propTypes = {
  actions: PropTypes.shape({
    logOut: PropTypes.func.isRequired,
    deleteAccount: PropTypes.func.isRequired
  }).isRequired,
  onSelect: PropTypes.func.isRequired,
  account: PropTypes.instanceOf(Object).isRequired,
  activeAccount: PropTypes.instanceOf(Object),
  intl: PropTypes.instanceOf(Object).isRequired
};

class SignUp extends Component {
  constructor(props) {
    super(props);

    this.state = {
      processing: false,
      username: '',
      email: '',
      refCode: '',
      done: false
    };
  }

  doSignUp = () => {
    const { username, email, refCode } = this.state;
    this.setState({ processing: true });

    return signUpFn(username, email, refCode)
      .then(resp => {
        this.setState({ done: true });
        return resp;
      })
      .catch(err => {
        if (err.response && err.response.data && err.response.data.message) {
          message.error(err.response.data.message);
        }
      })
      .finally(() => {
        this.setState({ processing: false });
      });
  };

  usernameChanged = e => {
    this.setState({ username: e.target.value });
  };

  emailChanged = e => {
    this.setState({ email: e.target.value });
  };

  refCodeChanged = e => {
    this.setState({ refCode: e.target.value });
  };

  login = () => {
    const { backFn } = this.props;
    backFn();
  };

  render() {
    const { intl } = this.props;
    const { processing, username, email, refCode, done } = this.state;

    return (
      <div className="sign-up-dialog-content">
        <div className="dialog-header">
          <div className="logo">
            <img src={logo} alt="logo" />
          </div>
          <div className="dialog-header-text">
            <FormattedMessage id="sign-up.title" />
          </div>
        </div>
        <div className="dialog-form">
          {(() => {
            if (done) {
              return (
                <div className="sign-up-success">
                  <p>
                    <i className="mi">check</i>
                    <FormattedMessage
                      id="sign-up.success-message-1"
                      values={{ email }}
                    />
                  </p>
                  <p>
                    <FormattedMessage id="sign-up.success-message-2" />
                  </p>
                  <p>
                    <Button
                      size="large"
                      htmlType="button"
                      type="primary"
                      block
                      onClick={this.login}
                    >
                      <FormattedMessage id="sign-up.login-btn-label" />
                    </Button>
                  </p>
                </div>
              );
            }

            return (
              <Fragment>
                <p>
                  <Input
                    type="text"
                    autoFocus
                    value={username}
                    onChange={this.usernameChanged}
                    placeholder={intl.formatMessage({
                      id: 'sign-up.username-placeholder'
                    })}
                  />
                </p>
                <p>
                  <Input
                    type="email"
                    value={email}
                    onChange={this.emailChanged}
                    placeholder={intl.formatMessage({
                      id: 'sign-up.email-placeholder'
                    })}
                  />
                </p>
                <p>
                  <Input
                    type="text"
                    value={refCode}
                    onChange={this.refCodeChanged}
                    placeholder={intl.formatMessage({
                      id: 'sign-up.ref-code'
                    })}
                  />
                </p>
                <p>
                  <Button
                    size="large"
                    htmlType="button"
                    type="primary"
                    block
                    disabled={processing}
                    onClick={this.doSignUp}
                  >
                    <FormattedMessage id="sign-up.sign-up" />
                  </Button>
                </p>
                <p>
                  <FormattedMessage id="sign-up.login-1" />
                  &nbsp;
                  <a onClick={this.login}>
                    <FormattedMessage id="sign-up.login-2" />
                  </a>
                </p>
              </Fragment>
            );
          })()}
        </div>
      </div>
    );
  }
}

SignUp.propTypes = {
  intl: PropTypes.instanceOf(Object).isRequired,
  backFn: PropTypes.func.isRequired
};

class Login extends Component {
  constructor(props) {
    super(props);

    this.state = {
      processing: false,
      signUp: false
    };
  }

  afterLogin = username => {
    const ev = new CustomEvent('user-login', { detail: { username } });
    window.dispatchEvent(ev);

    usrActivity(username, 20);
  };

  switchToAccount = username => {
    const { actions, onSuccess } = this.props;
    actions.logIn(username);
    onSuccess(username);
    this.afterLogin(username);
  };

  doScLogin = async () => {
    let resp;
    try {
      const loginResp = await scLogin();
      resp = await scTokenRenew(loginResp.code);
    } catch (e) {
      return;
    }

    const { actions, onSuccess } = this.props;
    actions.addAccountSc(
      resp.username,
      resp.access_token,
      resp.refresh_token,
      resp.expires_in
    );
    actions.logIn(resp.username);
    onSuccess(resp.username);
    this.afterLogin(resp.username);
  };

  doLogin = async () => {
    const { intl } = this.props;

    const username = document.querySelector('#txt-username').value.trim();
    const code = document.querySelector('#txt-code').value.trim();

    // Form validation
    if (username === '' || code === '') {
      message.error(intl.formatMessage({ id: 'login.error-fields-required' }));
      return false;
    }

    // Warn user if entered a public key
    let isPubkey;
    try {
      PublicKey.fromString(code);
      isPubkey = true;
    } catch (e) {
      isPubkey = false;
    }

    if (isPubkey) {
      message.error(intl.formatMessage({ id: 'login.error-public-key' }));
      return false;
    }

    // True if the code entered is password else false
    // const codeIsPassword = !auth.isWif(code);
    let codeIsPassword = false;
    try {
      PrivateKey.fromString(code);
    } catch (e) {
      codeIsPassword = true;
    }

    let accounts;

    this.setState({ processing: true });

    try {
      accounts = await getAccounts([username]);
    } catch (err) {
      message.error(intl.formatMessage({ id: 'login.error-user-fetch' }));
      return;
    } finally {
      this.setState({ processing: false });
    }

    // User not found
    if (accounts && accounts.length === 0) {
      message.error(intl.formatMessage({ id: 'login.error-user-not-found' }));
      return false;
    }

    // Account data
    const account = accounts[0];

    // Public keys of user
    const userPublicKeys = {
      active: account.active.key_auths.map(x => x[0]),
      memo: [account.memo_key],
      owner: account.owner.key_auths.map(x => x[0]),
      posting: account.posting.key_auths.map(x => x[0])
    };

    let loginFlag = false;
    const resultKeys = { active: null, memo: null, owner: null, posting: null };

    if (codeIsPassword) {
      // With master password

      // Get all private keys by username and password
      const userPrivateKeys = {};
      ['owner', 'active', 'posting', 'memo'].forEach(r => {
        const k = PrivateKey.fromLogin(account.name, code, r);

        userPrivateKeys[r] = k.toString();
        userPrivateKeys[`${r}Pubkey`] = k.createPublic().toString();
      });

      Object.keys(userPublicKeys).map(k => {
        const v = userPublicKeys[k];
        const v2 = userPrivateKeys[`${k}Pubkey`];

        if (v.includes(v2)) {
          resultKeys[k] = userPrivateKeys[k];
          loginFlag = true;
        }

        return null;
      });
    } else {
      // With wif
      const publicWif = PrivateKey.fromString(code)
        .createPublic()
        .toString();

      Object.keys(userPublicKeys).map(k => {
        const v = userPublicKeys[k];
        if (v.includes(publicWif)) {
          loginFlag = true;
          resultKeys[k] = code;
        }

        return null;
      });
    }

    if (!loginFlag) {
      message.error(intl.formatMessage({ id: 'login.error-authenticate' }));
      return false;
    }

    const { actions, onSuccess } = this.props;
    actions.addAccount(username, resultKeys);
    actions.logIn(username);
    onSuccess(username);
    this.afterLogin(username);
  };

  toggleSignUp = () => {
    const { signUp, processing } = this.state;
    if (processing) {
      return;
    }
    this.setState({ signUp: !signUp });
  };

  render() {
    const { accounts, loginMsg, intl } = this.props;
    const { processing, signUp } = this.state;

    if (signUp) {
      return <SignUp {...this.props} backFn={this.toggleSignUp} />;
    }

    return (
      <div className="login-dialog-content">
        <div className="dialog-header">
          <div className="logo">
            <img src={logo} alt="logo" />
          </div>
          <div className="dialog-header-text">
            <FormattedMessage id="login.title" />
          </div>
        </div>

        {loginMsg && (
          <div className="login-msg">
            <Alert message={loginMsg} showIcon />
          </div>
        )}

        {accounts.length > 0 && (
          <div className="account-list">
            <div className="account-list-header">
              <FormattedMessage id="login.accounts-title" />
            </div>
            <div className="account-list-body">
              {accounts.map(account => (
                <AccountItem
                  key={account.username}
                  account={account}
                  onSelect={this.switchToAccount}
                  {...this.props}
                />
              ))}
            </div>
          </div>
        )}

        {accounts.length > 0 && (
          <Divider>
            <FormattedMessage id="login.divider-text" />
          </Divider>
        )}

        <div className="dialog-form">
          <p className="form-text">
            <FormattedMessage id="login.traditional-login-desc" />
          </p>
          <p>
            <Input
              type="text"
              autoFocus
              id="txt-username"
              placeholder={intl.formatMessage({
                id: 'login.username-placeholder'
              })}
            />
          </p>
          <p>
            <Input
              type="password"
              id="txt-code"
              placeholder={intl.formatMessage({
                id: 'login.password-placeholder'
              })}
            />
          </p>
          <p>
            <Button
              size="large"
              htmlType="button"
              type="primary"
              block
              disabled={processing}
              onClick={this.doLogin}
            >
              <FormattedMessage id="login.login" />
            </Button>
          </p>
          <p>
            <FormattedMessage id="login.create-account-text-1" />
            &nbsp;
            <a onClick={this.toggleSignUp}>
              <FormattedMessage id="login.create-account-text-2" />
            </a>
          </p>
        </div>
        <Divider>
          <FormattedMessage id="login.divider-text" />
        </Divider>
        <div className="login-sc" role="none" onClick={this.doScLogin}>
          <p className="login-title">
            <FormattedMessage id="login.login-with-sc" />
          </p>
          <div className="logo">
            <img src={scLogo} alt="Steem Connect" />
          </div>
          <div className="sc-note">
            <FormattedMessage id="login.sc-note" />
          </div>
        </div>
      </div>
    );
  }
}

Login.defaultProps = {
  loginMsg: null,
  accounts: []
};

Login.propTypes = {
  actions: PropTypes.shape({
    addAccount: PropTypes.func.isRequired,
    addAccountSc: PropTypes.func.isRequired,
    logIn: PropTypes.func.isRequired,
    updateActiveAccount: PropTypes.func.isRequired
  }).isRequired,
  onSuccess: PropTypes.func.isRequired,
  accounts: PropTypes.arrayOf(PropTypes.object),
  loginMsg: PropTypes.element,
  intl: PropTypes.instanceOf(Object).isRequired
};

export default Login;
