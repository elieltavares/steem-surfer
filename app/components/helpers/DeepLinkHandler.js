import React from 'react';
import PropTypes from 'prop-types';

import filters from '../../constants/filters.json';

import { makePath as makePathTag } from './TagLink';
import { makePath as makePathAccount } from './AccountLink';
import { makePath as makePathEntry } from './EntryLink';

export const deepLink2Obj = url => {
  let urlPart = url.split('://')[1];

  // remove last char if /
  if (urlPart.endsWith('/')) {
    urlPart = urlPart.substring(0, urlPart.length - 1);
  }

  const parts = urlPart.split('/');

  // witnesses
  if (parts.length === 1 && parts[0] === 'witnesses') {
    return { type: 'witnesses' };
  }

  // sps
  if (parts.length === 1 && parts[0] === 'sps') {
    return { type: 'sps' };
  }

  // filter
  if (parts.length === 1 && filters.includes(parts[0])) {
    return { type: 'filter', filter: parts[0] };
  }

  // filter with tag
  if (parts.length === 2 && filters.includes(parts[0])) {
    return { type: 'filter-tag', filter: parts[0], tag: parts[1] };
  }

  // account
  if (parts.length === 1 && parts[0].startsWith('@')) {
    return { type: 'account', account: parts[0].replace('@', '') };
  }

  // post without tag
  if (parts.length === 2 && parts[0].startsWith('@')) {
    return {
      type: 'post',
      cat: 'esteem',
      author: parts[0].replace('@', ''),
      permlink: parts[1]
    };
  }

  // post with tag
  if (parts.length === 3 && parts[1].startsWith('@')) {
    return {
      type: 'post',
      cat: parts[0],
      author: parts[1].replace('@', ''),
      permlink: parts[2]
    };
  }
};

class DeepLinkHandler extends React.Component {
  componentDidMount() {
    window.addEventListener('deep-link', this.handler);
  }

  componentWillUnmount() {
    window.removeEventListener('deep-link', this.handler);
  }

  handler = e => {
    const { history } = this.props;
    const { url } = e.detail;
    const obj = deepLink2Obj(url);

    let path;

    switch (obj.type) {
      case 'filter':
        path = `/${obj.filter}`;
        break;
      case 'filter-tag':
        path = makePathTag(obj.filter, obj.tag);
        break;
      case 'account':
        path = makePathAccount(obj.account);
        break;
      case 'post':
        path = makePathEntry(obj.cat, obj.author, obj.permlink);
        break;
      case 'sps':
        path = `/sps`;
        break;
      case 'witnesses':
        path = `/witnesses`;
        break;
      default:
        path = null;
        break;
    }

    if (path) {
      history.push(path);
    }
  };

  render() {
    return null;
  }
}

DeepLinkHandler.propTypes = {
  history: PropTypes.shape({}).isRequired
};

export default DeepLinkHandler;
