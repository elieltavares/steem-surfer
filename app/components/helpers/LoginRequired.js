import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';

import { Modal } from 'antd';
import { FormattedMessage, FormattedHTMLMessage } from 'react-intl';

import Login from '../dialogs/Login';

export const checkLogin = (activeAccount, requiredKeys) => {
  // Steem connect login.
  if (activeAccount && activeAccount.type === 'sc') {
    return true;
  }

  // Traditional login. Check all required keys exists in user keys.
  if (
    activeAccount &&
    activeAccount.type === 's' &&
    requiredKeys.every(e => activeAccount.keys[e])
  ) {
    return true;
  }

  return false;
};

class LoginRequired extends Component {
  constructor(props) {
    super(props);

    this.state = {
      loginModalVisible: false
    };
  }

  showLoginModal = () => {
    this.setState({
      loginModalVisible: true
    });

    setTimeout(() => {
      const { onDialogOpen } = this.props;
      onDialogOpen();
    }, 200);
  };

  onLoginModalCancel = () => {
    this.setState({
      loginModalVisible: false
    });
  };

  onLoginSuccess = () => {
    this.setState({
      loginModalVisible: false
    });

    setTimeout(() => {
      if (!this.check()) {
        this.showLoginModal();
      }
    }, 400);
  };

  afterModalClose = () => {
    const { onDialogClose } = this.props;
    onDialogClose();
  };

  check = () => {
    const { activeAccount, requiredKeys } = this.props;

    return checkLogin(activeAccount, requiredKeys);
  };

  render() {
    const { children } = this.props;

    if (this.check()) {
      return children;
    }

    const { activeAccount, requiredKeys } = this.props;
    const { loginModalVisible } = this.state;

    // Default message
    let loginMsg = <FormattedMessage id="login-required.default" />;

    if (activeAccount) {
      // More specific, key based message
      loginMsg =
        requiredKeys.length > 1 ? (
          <FormattedHTMLMessage
            id="login-required.keys-required"
            values={{ keys: requiredKeys }}
          />
        ) : (
          <FormattedHTMLMessage
            id="login-required.key-required"
            values={{ key: requiredKeys }}
          />
        );
    }

    // Clone child element rewriting onclick event
    const newChild = React.cloneElement(children, {
      onClick: () => {
        this.showLoginModal();
      }
    });

    return (
      <Fragment>
        {newChild}

        <Modal
          visible={loginModalVisible}
          onCancel={this.onLoginModalCancel}
          footer={false}
          width="500px"
          closable={false}
          destroyOnClose
          centered
          afterClose={this.afterModalClose}
        >
          <Login
            {...this.props}
            onSuccess={this.onLoginSuccess}
            loginMsg={loginMsg}
          />
        </Modal>
      </Fragment>
    );
  }
}

LoginRequired.defaultProps = {
  activeAccount: null,
  onDialogOpen: () => {},
  onDialogClose: () => {},
  requiredKeys: []
};

LoginRequired.propTypes = {
  children: PropTypes.element.isRequired,
  activeAccount: PropTypes.instanceOf(Object),
  requiredKeys: PropTypes.arrayOf(PropTypes.string),
  onDialogOpen: PropTypes.func,
  onDialogClose: PropTypes.func
};

export default LoginRequired;
