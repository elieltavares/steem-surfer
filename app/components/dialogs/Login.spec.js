/* eslint-disable */

import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Login from './Login';
import { mountWithIntl } from '../../helpers/intl-test';

Enzyme.configure({ adapter: new Adapter() });

describe('Login', () => {
  it('(1) Default Render', () => {
    const props = {
      actions: {
        addAccount: () => {},
        addAccountSc: () => {},
        logIn: () => {},
        logOut: () => {},
        deleteAccount: () => {},
        updateActiveAccount: () => {}
      },
      onSuccess: () => {},
      accounts: []
    };

    const component = mountWithIntl(<Login {...props} />);
    expect(component.html()).toMatchSnapshot();
  });

  it('(3) Render with account list', () => {
    const props = {
      actions: {
        addAccount: () => {},
        addAccountSc: () => {},
        logIn: () => {},
        logOut: () => {},
        deleteAccount: () => {},
        updateActiveAccount: () => {}
      },
      onSuccess: () => {},
      accounts: [
        {
          username: 'foo'
        },
        {
          username: 'bar'
        }
      ]
    };

    const component = mountWithIntl(<Login {...props} />);
    expect(component.html()).toMatchSnapshot();
  });
});
