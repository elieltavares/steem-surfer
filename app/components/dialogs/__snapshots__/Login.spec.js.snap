// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login (1) Default Render 1`] = `"<div class=\\"login-dialog-content\\"><div class=\\"login-header\\"><div class=\\"logo\\"><img src=\\"test-file-stub\\" alt=\\"logo\\"></div><div class=\\"login-header-text\\"><span>Welcome back!</span></div></div><div class=\\"login-form\\"><p class=\\"form-text\\"><span>Sign in with your username and password to continue</span></p><p><input type=\\"text\\" id=\\"txt-username\\" placeholder=\\"username\\" class=\\"ant-input\\" value=\\"\\"></p><p><input type=\\"password\\" id=\\"txt-code\\" placeholder=\\"password or WIF\\" class=\\"ant-input\\" value=\\"\\"></p><p><button type=\\"button\\" class=\\"ant-btn ant-btn-primary ant-btn-lg ant-btn-block\\"><span>Login</span></button></p><p><span>Don't have an account? <a href=\\"https://signup.steemit.com/?ref=esteemapp\\" target=\\"_external\\">Sign up now!</a></span></p></div><div class=\\"ant-divider ant-divider-horizontal ant-divider-with-text\\"><span class=\\"ant-divider-inner-text\\"><span>OR</span></span></div><div class=\\"login-sc\\" role=\\"none\\"><p class=\\"login-title\\"><span>Login with</span></p><div class=\\"logo\\"><img src=\\"test-file-stub\\" alt=\\"Steem Connect\\"></div><div class=\\"sc-note\\"><span>Note: Steemconnect may charge some fees from your reward transactions</span></div></div></div>"`;

exports[`Login (3) Render with account list 1`] = `"<div class=\\"login-dialog-content\\"><div class=\\"login-header\\"><div class=\\"logo\\"><img src=\\"test-file-stub\\" alt=\\"logo\\"></div><div class=\\"login-header-text\\"><span>Welcome back!</span></div></div><div class=\\"account-list\\"><div class=\\"account-list-header\\"><span>Login As</span></div><div class=\\"account-list-body\\"><div class=\\"account-list-item \\" role=\\"none\\"><span role=\\"none\\" class=\\"user-avatar normal\\" style=\\"background-image: url(https://avatars.esteem.app/u/foo/avatar/medium);\\"></span> <span class=\\"username\\">@foo</span><div class=\\"space\\"></div><div role=\\"none\\" class=\\"btn-delete\\"><i class=\\"mi tooltip-handler\\">delete_forever</i></div></div><div class=\\"account-list-item \\" role=\\"none\\"><span role=\\"none\\" class=\\"user-avatar normal\\" style=\\"background-image: url(https://avatars.esteem.app/u/bar/avatar/medium);\\"></span> <span class=\\"username\\">@bar</span><div class=\\"space\\"></div><div role=\\"none\\" class=\\"btn-delete\\"><i class=\\"mi tooltip-handler\\">delete_forever</i></div></div></div></div><div class=\\"ant-divider ant-divider-horizontal ant-divider-with-text\\"><span class=\\"ant-divider-inner-text\\"><span>OR</span></span></div><div class=\\"login-form\\"><p class=\\"form-text\\"><span>Sign in with your username and password to continue</span></p><p><input type=\\"text\\" id=\\"txt-username\\" placeholder=\\"username\\" class=\\"ant-input\\" value=\\"\\"></p><p><input type=\\"password\\" id=\\"txt-code\\" placeholder=\\"password or WIF\\" class=\\"ant-input\\" value=\\"\\"></p><p><button type=\\"button\\" class=\\"ant-btn ant-btn-primary ant-btn-lg ant-btn-block\\"><span>Login</span></button></p><p><span>Don't have an account? <a href=\\"https://signup.steemit.com/?ref=esteemapp\\" target=\\"_external\\">Sign up now!</a></span></p></div><div class=\\"ant-divider ant-divider-horizontal ant-divider-with-text\\"><span class=\\"ant-divider-inner-text\\"><span>OR</span></span></div><div class=\\"login-sc\\" role=\\"none\\"><p class=\\"login-title\\"><span>Login with</span></p><div class=\\"logo\\"><img src=\\"test-file-stub\\" alt=\\"Steem Connect\\"></div><div class=\\"sc-note\\"><span>Note: Steemconnect may charge some fees from your reward transactions</span></div></div></div>"`;
